### PR TITLE
feat: track policy training compute throughput

### LINKED ISSUE #59: MFU metric
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

Provide a code snippet on how new APIs/changes would be used by others.

**Describe alternatives you've considered**

**Additional context**

Add any other context or screenshots about the feature request here

### PR DESCRIPTION
**Pull request by ybgao-nvidia**
# What does this PR do ?

Allows tracking of FLOPS (floating point operations per second) for policy training, which can be used to estimate MFU. 

This PR adds:
1. The `FLOPCounter` class: maintains a sum of FLOPs incurred through model training at a batch granularity. After processing each batch, the trainer passes in a tuple `(batch size, sequence length)` into the counter, which computes and accumulates the total FLOPs. 
2. Invocation of the tracker in the DTensor worker. 
3. Additions in the GRPO, DPO and SFT training scripts to track the time consumed and compute the compute throughput (FLOPs per second) for each training iteration. 
4. Training FLOPs calculation for Qwen2 (previously only Qwen3 is included)
5. MFU metric logging in console and wandb (as `train/train_fp_utilization` if enabled)

Limitations:
1. We have currently only implemented support for DTensor worker.
2. The conversion from HF's `PretrainedConfig` to `FLOPSConfig` only supports `Qwen2Config` and `LlamaConfig`

# Issues
This PR resolves #59. 

# Usage
When training with a supported model, we will see the FLOPS in addition to the training time upon completion of each epoch as follows:

```
📊 Training Results:
  • Loss: 0.0121
  • Avg Reward: 0.3809
  • Mean Generation Length: 285.2734
  • Training FLOPS: 774.91 TFLOPS (96.86 TFLOPS per rank)
  • Training Model Floating Point Utilization: 9.79%
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information

### CODE REVIEW

File: nemo_rl/models/policy/lm_policy.py [hunk 1]
```diff
@@ -325,6 +325,10 @@ def train(
             "grad_norm": results[0]["grad_norm"],
         }
 
+        if all("total_flops" in r and r["total_flops"] is not None for r in results):
```
Comment (parthchadha) [inline nemo_rl/models/policy/lm_policy.py]: Adding a comment to ensure we add logging of MFU (and not just flops)

Comment (ybgao-nvidia) [reply nemo_rl/models/policy/lm_policy.py]: I have added logging of MFU in both console and wandb.

File: nemo_rl/utils/flops_formulas.py [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+from typing import List, Optional, Union
+
+# lifted from NeMo/nemo/collections/common/parts/perf_metrics_utils.py
+LLM_VOCAB_SIZE_MAP = {
```
Comment (jiemingz) [inline nemo_rl/utils/flops_formulas.py]: is it possible to get the vocab size from the model rather than hardcoding it?

File: nemo_rl/utils/flops_formulas.py [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+}
+
+
+# lifted from NeMo/nemo/utils/flops_formulas.py
+@dataclass
+class FLOPSConfig:
+    """Contains the model hparams needed for FLOPS computations."""
+
+    gbs: int
+    enc_seq_len: Optional[int] = None
```
Comment (jiemingz) [inline nemo_rl/utils/flops_formulas.py:25]: here the squence length is constant, but we have variable sequence lengths. Our effective MFU should be considering flops from only non padding tokens

File: nemo_rl/utils/flops_formulas.py [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+            )
+            # Logit component
+            + (vocab_size / (2 * num_layers * hidden_size))
+        )
+    )
+
+    return total_flops
+
+
+def clip_vit_l(config: FLOPSConfig):
```
Comment (jiemingz) [inline nemo_rl/utils/flops_formulas.py]: we dont support vit or neva models, should we remove the unused formulas?

File: nemo_rl/utils/flops_formulas.py (hunk 2 above)
Comment (ybgao-nvidia) [reply nemo_rl/utils/flops_formulas.py]: Yes, now the vocab size is read from model config.

File: nemo_rl/utils/flops_formulas.py (hunk 4 above)
Comment (ybgao-nvidia) [reply nemo_rl/utils/flops_formulas.py]: Removed.

File: nemo_rl/models/policy/dtensor_policy_worker.py [hunk 5]
```diff
@@ -175,6 +176,12 @@ def __init__(
             ),  # due to https://github.com/huggingface/transformers/issues/38002
         )
 
+        try:
+            self.flops_tracker = FLOPTracker.from_config(model_name, model_config)
```
Comment (jiemingz) [inline nemo_rl/models/policy/dtensor_policy_worker.py]: would it be possible to move this logic into lm_policy? that way it isnt dtensor specific and works for all training backend

Comment (ybgao-nvidia) [reply nemo_rl/models/policy/dtensor_policy_worker.py]: Yes, although the actual FLOPs tracking logic still has to be backend-specific. For example, for the dtensor worker we need to:

https://github.com/ybgao-nvidia/RL/blob/9b2122875554654d1f3b33f1c95c899b6166e5b1/nemo_rl/models/policy/dtensor_policy_worker.py#L519-L521

Comment (ybgao-nvidia) [reply nemo_rl/models/policy/dtensor_policy_worker.py]: It is also easier to guarantee correctness since the microbatch iterator is initialized in the policy worker, and we call `track()` right before we process a microbatch.

Comment (jiemingz) [reply nemo_rl/models/policy/dtensor_policy_worker.py]: flops_tracker just needs the sequence lengths per rank right?
could we get the input_lengths at the global batch instead of the iterator?
for instance i think you could get the sequence lengths here:
https://github.com/ybgao-nvidia/RL/blob/9b2122875554654d1f3b33f1c95c899b6166e5b1/nemo_rl/models/policy/lm_policy.py#L173

```
for data_shard in sharded_data:
     input_lengths = data_shard['input_lengths']
```
   
 you can also calculate the padding amount as its is stored with: https://github.com/ybgao-nvidia/RL/blob/9b2122875554654d1f3b33f1c95c899b6166e5b1/nemo_rl/distributed/batched_data_dict.py#L456
 also do you account for context parallelism? CP divides the number of tokens per rank by CP

Comment (jiemingz) [reply nemo_rl/models/policy/dtensor_policy_worker.py]: BTW I think
```
if self.flops_tracker:
         for l in input_lengths.tolist():
                      self.flops_tracker.track(1, l)
```
would trigger a host sync every time its run which is undesirable in the training loop (though we currently have many unnecessary host syncs that we need to remove so this might not matter currently)

Comment (ybgao-nvidia) [reply nemo_rl/models/policy/dtensor_policy_worker.py]: Thanks for the input; I've addressed these in the latest commit.

File: nemo_rl/models/policy/dtensor_policy_worker.py [hunk 6]
```diff
@@ -709,6 +709,8 @@ def train(
                 "global_loss": global_loss.cpu(),
                 "grad_norm": grad_norm,
                 "rank": torch.distributed.get_rank(),
+                "gpu_name": torch.cuda.get_device_name(),
+                "model_dtype": self.dtype,
```
Comment (jiemingz) [inline nemo_rl/models/policy/dtensor_policy_worker.py:865]: would this work with the megatron backend?
i suppose if you were move these changes into lm_policy; you can get `self.dtype` from self.cfg['precision']
https://github.com/NVIDIA-NeMo/RL/blob/3138e00aca5081760aeef479441fab3a868d1f2a/nemo_rl/models/policy/dtensor_policy_worker.py#L164
and the gpu name is the same everywhere

File: nemo_rl/models/policy/lm_policy.py [hunk 7]
```diff
@@ -330,6 +350,18 @@ def train(
...
+            aggregated_results["total_flops"] = self.flops_tracker.total_flops
+            aggregated_results["num_ranks"] = len(results)
+
+            try:
+                aggregated_results["theoretical_flops"] = sum(
+                    get_theoretical_flops(r["gpu_name"], r["model_dtype"])
+                    for r in results
+                )
+            except Exception as e:
+                print(f"Error getting theoretical flops: {e}")
```
Comment (jiemingz) [inline nemo_rl/models/policy/lm_policy.py]: nit: in the case where flop calculator encounters an unsupported config, it might be nice to log that once at setup or else every train step it will print an error message and pollute the logs

**Comment by jiemingz** [comment]
could you also make sure your changes are covered by tests?
https://github.com/NVIDIA-NeMo/RL/blob/a607b417c0b5fbdd677a95c63fe37ef1044862e8/tests/unit/models/generation/test_vllm_generation.py#L345

File: nemo_rl/models/policy/dtensor_policy_worker.py (hunk 6 above)
Comment (ybgao-nvidia) [reply nemo_rl/models/policy/dtensor_policy_worker.py:865]: This should also work with megatron backend since it's set here:
https://github.com/NVIDIA-NeMo/RL/blob/3138e00aca5081760aeef479441fab3a868d1f2a/nemo_rl/models/policy/megatron_policy_worker.py#L359-L364

File: nemo_rl/models/policy/lm_policy.py (hunk 7 above)
Comment (ybgao-nvidia) [reply nemo_rl/models/policy/lm_policy.py]: Updated with `warnings.warn()` which will only print the warning the first time.

**Review by jiemingz — DISMISSED** [review]

**Review by jiemingz — APPROVED** [review]

**Review by parthchadha — DISMISSED** [review]

**Review by jiemingz — DISMISSED** [review]

**Review by parthchadha — DISMISSED** [review]

**Review by jiemingz — DISMISSED** [review]

**Review by jiemingz — APPROVED** [review]

**Review by terrykong — APPROVED** [review]
